File: backend-order/models/order.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type OrderProduct struct {
	ID    string  `json:"id" bson:"id"`
	Name  string  `json:"name" bson:"name"`
	Price float64 `json:"price" bson:"price"`
}

type TimelineEvent struct {
	Name      string    `json:"name" bson:"name"`
	Timestamp time.Time `json:"timestamp" bson:"timestamp"`
}

type Order struct {
	ID          primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	CustomerID  string             `json:"customer_id" bson:"customer_id"`
	Product     OrderProduct       `json:"product" bson:"product"`
	Quantity    int                `json:"quantity" bson:"quantity"`
	TotalAmount float64            `json:"total_amount" bson:"total_amount"`
	Status      string             `json:"status" bson:"status"`
	PaymentID   string             `json:"payment_id,omitempty" bson:"payment_id,omitempty"`
	CreatedAt   time.Time          `json:"created_at" bson:"created_at"`
	UpdatedAt   time.Time          `json:"updated_at" bson:"updated_at"`
	Timeline    []TimelineEvent    `json:"timeline" bson:"timeline"`
}

const (
	OrderStatusCreated   = "Created"
	OrderStatusConfirmed = "Confirmed"
	OrderStatusCancelled = "Cancelled"
	OrderStatusDelivered = "Delivered"
)

File: frontend/src/api/Order.ts
import { API_ORDER_URL, API_PAYMENT_URL } from './config';

export interface OrderProduct {
  id: string;
  name: string;
  price: number;
}

export interface TimelineEvent {
  name: string;
  timestamp: string;
}

export interface Order {
  id: string;
  customer_id: string;
  product: OrderProduct;
  quantity: number;
  total_amount: number;
  status: 'Created' | 'Confirmed' | 'Delivered' | 'Cancelled';
  payment_id?: string;
  created_at: string;
  updated_at: string;
  timeline: TimelineEvent[];
}

export interface CreateOrderRequest {
  customer_id: string;
  product_id: string;
  quantity: number;
}

export const createOrder = async (orderData: CreateOrderRequest): Promise<Order> => {
  const token = localStorage.getItem('token');
  if (!token) {
    throw new Error('No authentication token found');
  }

  const response = await fetch(`${API_ORDER_URL}/orders`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': token,
    },
    body: JSON.stringify(orderData),
  });

  if (!response.ok) {
    throw new Error('Failed to create order');
  }

  const data = await response.json();
  return data.order;
};

export const getOrders = async (): Promise<Order[]> => {
  const token = localStorage.getItem('token');
  if (!token) {
    throw new Error('No authentication token found');
  }

  const response = await fetch(`${API_ORDER_URL}/orders`, {
    headers: {
      'Authorization': token,
    },
  });

  if (!response.ok) {
    throw new Error('Failed to fetch orders');
  }

  return response.json();
};

export const cancelOrder = async (orderId: string): Promise<void> => {
  const token = localStorage.getItem('token');
  if (!token) {
    throw new Error('No authentication token found');
  }

  const response = await fetch(`${API_ORDER_URL}/orders/${orderId}/cancel`, {
    method: 'POST',
    headers: {
      'Authorization': token,
    },
  });

  if (!response.ok) {
    throw new Error('Failed to cancel order');
  }
};

export const initiatePayment = async (orderId: string, amount: number): Promise<void> => {
  const token = localStorage.getItem('token');
  if (!token) {
    throw new Error('No authentication token found');
  }

  const response = await fetch(`${API_PAYMENT_URL}/payments`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': token,
    },
    body: JSON.stringify({ order_id: orderId, amount }),
  });

  if (!response.ok) {
    throw new Error('Failed to initiate payment');
  }
};
